mypy/semanal.py

```diff
@@ -719,9 +719,54 @@ def analyze_class_body(self, defn: ClassDef) -> Iterator[bool]:
                 yield True
                 self.calculate_abstract_status(defn.info)
                 self.setup_type_promotion(defn)
-
+                self.apply_class_plugin_hooks(defn)
                 self.leave_class()
 
+    def apply_class_plugin_hooks(self, defn: ClassDef) -> None:
+        """Apply a plugin hook that may infer a more precise definition for a class."""
+
+        def get_fullname(expr: Expression) -> Optional[str]:
+            # We support foo.bar(...), foo.bar, and bar.
+            if isinstance(expr, CallExpr):
+                if isinstance(expr.callee, RefExpr):
+                    return expr.callee.fullname
+            elif isinstance(expr, MemberExpr):
+                return get_member_expr_fullname(expr)
+            elif isinstance(expr, NameExpr):
+                return expr.fullname
+            return None
+
+        for decorator in defn.decorators:
+            decorator_name = get_fullname(decorator)
+            if decorator_name:
+                hook = self.plugin.get_class_decorator_hook(decorator_name)
+                if hook:
+                    hook(ClassDefContext(defn, decorator, self))
+
+        if defn.metaclass:
+            metaclass_name = get_fullname(defn.metaclass)
+            if metaclass_name:
+                hook = self.plugin.get_class_metaclass_hook(metaclass_name)
+                if hook:
+                    hook(ClassDefContext(defn, defn.metaclass, self))
+
+        for base_expr in defn.base_type_exprs:
+            try:
+                base = expr_to_unanalyzed_type(base_expr)
+            except TypeTranslationError:
+                continue  # This will be reported later
+            if not isinstance(base, UnboundType):
+                continue
+            sym = self.lookup_qualified(base.name, base)
+            if sym is None or sym.node is None:
+                continue
+            base_name = sym.node.fullname()
```

Comment:
Why this is more complex than in other cases? I would expect you can just extract the name using ``get_full_name`` (if you would add one more case for ``IndexExpr``).

Reply:
Because of this:
```
    # Base class expressions (not semantically analyzed -- can be arbitrary expressions)
    base_type_exprs = None  # type: List[Expression]
```

If I ask for the `expr.base.fullname` of `class Foo(Mapping[str, str])` I get `Mapping` instead of `typing.Mapping`